# 线程中的异常处理

## 处理线程中的异常

- setUncaughtExceptionHandle()：给指定线程对象设置异常处理器。
- setDefaultUncaughtExceptionHandle()：对一个线程类的所有对象设置异常处理器。

**测试 setUncaughtExceptionHandle：**

```java
package ch7.simpledateformat.controller;

public class UncaughtExceptionHandleTest {
    public static void main(String[] args) {
        ErrorThread thread1 = new ErrorThread();
        // 因为是给指定线程对象设置异常处理器，所以是线程对象在调用
        thread1.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
            @Override
            public void uncaughtException(Thread t, Throwable e) {
                System.out.println("线程 " + t.getName() + " 出现了异常");
                e.printStackTrace();
            }
        });
        thread1.start();

        ErrorThread thread2 = new ErrorThread();
        thread2.start();
    }
}
```

**测试 setDefaultUncaughtExceptionHandle：**

```java
package ch7.simpledateformat.controller;

public class DefaultUncaughtExceptionHandleTest {
    public static void main(String[] args) {
        // 因为是对一个线程类的所有对象设置异常处理器，所以是线程类在调用
        ErrorThread.setDefaultUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
            @Override
            public void uncaughtException(Thread t, Throwable e) {
                System.out.println("线程 " + t.getName() + " 出现了异常");
                e.printStackTrace();
            }
        });

        ErrorThread thread1 = new ErrorThread();
        thread1.start();
        ErrorThread thread2 = new ErrorThread();
        thread2.start();
    }
}
```

*一个会抛出异常的线程类：*

```java
package ch7.simpledateformat.controller;

public class ErrorThread extends Thread {
    @Override
    public void run() {
        String username = null;
        System.out.println(username.hashCode());
    }
}
```

## 处理线程组中的异常

**应用：实现一个线程组中的一个线程出现了异常后组内线程全部停止**

**方法：**新写一个 ThreadGroup 的子类，覆盖 ThreadGroup 中的 public void uncaughtException(Thread t, Throwable e) 方法：

```java
package ch7.simpledateformat.stopgroup;

public class MyThreadGroup extends ThreadGroup {
    public MyThreadGroup(String name) {
        super(name);
    }

    // 在重写的uncaughtException方法中处理组内线程的中断操作，这样组内有一个线程跑出来异常，
    // 该线程组会调用this.interrupt()，就中断了自己组内的所有线程
    @Override
    public void uncaughtException(Thread t, Throwable e) {
        super.uncaughtException(t, e);
        this.interrupt();
    }
}
```

**注意：**使用这种方法，组内的每个线程对象中的 run() 方法中不能有 catch 语句，如果有则不会执行 public void uncaughtException(Thread t, Throwable e) 方法。